src/backend/cpu/sift.cpp

```diff
@@ -0,0 +1,1046 @@
+            float xbin = x_rot + d/2 - 0.5f;
+            float ybin = y_rot + d/2 - 0.5f;
+
+            if (ybin > -1.0f && ybin < d && xbin > -1.0f && xbin < d &&
+                y > 0 && y < ddims[0] - 1 && x > 0 && x < ddims[1] - 1) {
+                float dx = img_ptr[(x+1)*ddims[0]+y] - img_ptr[(x-1)*ddims[0]+y];
+                float dy = img_ptr[x*ddims[0]+(y-1)] - img_ptr[x*ddims[0]+(y+1)];
+
+                float grad_mag = sqrt(dx*dx + dy*dy);
+                float grad_ori = atan2(dy, dx) - ori;
+                while (grad_ori < 0.0f)
+                    grad_ori += PI_VAL*2;
+                while (grad_ori >= PI_VAL*2)
+                    grad_ori -= PI_VAL*2;
+
+                float w = exp(-(x_rot*x_rot + y_rot*y_rot) / exp_denom);
+                float obin = grad_ori * bins_per_rad;
+                float mag = grad_mag*w;
+
+                int x0 = floor(xbin);
+                int y0 = floor(ybin);
+                int o0 = floor(obin);
+                xbin -= x0;
+                ybin -= y0;
+                obin -= o0;
+
+                for (int yl = 0; yl <= 1; yl++) {
+                    int yb = y0 + yl;
+                    if (yb >= 0 && yb < d) {
+	                    float v_y = mag * ((yl == 0) ? 1.0f - ybin : ybin);
+	                    for (int xl = 0; xl <= 1; xl++) {
+	                        int xb = x0 + xl;
+	                        if (xb >= 0 && xb < d) {
+		                        float v_x = v_y * ((xl == 0) ? 1.0f - xbin : xbin);
+		                        for (int ol = 0; ol <= 1; ol++) {
+		                            int ob = (o0 + ol) % n;
+		                            float v_o = v_x * ((ol == 0) ? 1.0f - obin : obin);
+		                            desc[(yb*d + xb)*n + ob] += v_o;
+		                        }
+		                    }
+	                    }
+	                }
+                }
+            }
+        }
+
+        normalizeDesc(desc, histlen);
+
+        for (int i = 0; i < d*d*n; i++)
+            desc[i] = min(desc[i], DescrMagThr);
+
+        normalizeDesc(desc, histlen);
+
+        // Calculate final descriptor values
+        for (int k = 0; k < d*d*n; k++) {
+            desc_out[f*desc_len+k] = round(min(255.f, desc[k] * IntDescrFctr));
+        }
+    }
+}
+
+template<typename T, typename convAccT>
+Array<T> createInitialImage(
+    const Array<T>& img,
+    const float init_sigma,
+    const bool double_input)
+{
+    af::dim4 idims = img.dims();
+
+    Array<T> init_img = createEmptyArray<T>(af::dim4());
+
+    float s = (double_input) ? sqrt(init_sigma * init_sigma - InitSigma * InitSigma * 4)
+                             : sqrt(init_sigma * init_sigma - InitSigma * InitSigma);
+
+    Array<T> filter = gauss_filter<T>(s);
+
+    if (double_input) {
+        Array<T> double_img = resize<T>(img, idims[0] * 2, idims[1] * 2, AF_INTERP_BILINEAR);
+        init_img = convolve2<T, convAccT, false>(double_img, filter, filter);
```

Comment:
It is dependent on the image size, for the first octaves, it may be larger, but there will be various octaves with small dimensions.